File: tvtwit/Util/UISearchBar+RACAdditions.m

#import <objc/runtime.h>
#import "UISearchBar+RACAdditions.h"
@implementation UISearchBar (RACAdditions)
- (RACSignal *)rac_textSignal {
    self.delegate = self;
    RACSignal *signal = objc_getAssociatedObject(self, _cmd);
    if (signal != nil) return signal;
    
    /* Create signal from selector */
    signal = [[self rac_signalForSelector:@selector(searchBar:textDidChange:)
                             fromProtocol:@protocol(UISearchBarDelegate)] map:^id(RACTuple *tuple) {
        return tuple.second;
    }];
    
    objc_setAssociatedObject(self, _cmd, signal, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
    return signal;
    return signal;
}
- (RACSignal *)rac_pressSearchSignal {
    self.delegate = self;
    RACSignal *signal = objc_getAssociatedObject(self, _cmd);
    if (signal != nil) return signal;
    
    /* Create signal from selector */
    signal = [[self rac_signalForSelector:@selector(searchBarSearchButtonClicked:)
                             fromProtocol:@protocol(UISearchBarDelegate)] map:^id(RACTuple *tuple) {
        return tuple.second;
    }];
    
    objc_setAssociatedObject(self, _cmd, signal, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
    return signal;
    return signal;
}
@end

File: tvtwit/Util/UISearchBar+RACAdditions.swift
//
//  UISearchBar+RACAdditions.swift
//  Hubtele
//
//  Created by NakamaTakaaki on 2016/02/03.
//  Copyright © 2016年 NakamaTakaaki. All rights reserved.
//

import Foundation
import UIKit
import ReactiveCocoa
extension UISearchBar:UISearchBarDelegate {
	func rac_textSignal() -> RACSignal {
		self.delegate = self;
        var signal:RACSignal? = objc_getAssociatedObject(self, __FUNCTION__) as? RACSignal
		if (signal != nil) {
			return signal!
		}
		
        signal = self.rac_signalForSelector("searchBar:textDidChange:", fromProtocol: UISearchBarDelegate.self).mapAs {(tuple:RACTuple) -> NSString in
			return tuple.second as! NSString
		}
        return signal!
	}
	func rac_pressSearchSignal() -> RACSignal {
		self.delegate = self
        var signal:RACSignal? = objc_getAssociatedObject(self, __FUNCTION__) as? RACSignal
        if (signal != nil) {
            return signal!
        }
		
        signal = self.rac_signalForSelector("searchBarSearchButtonClicked:", fromProtocol: UISearchBarDelegate.self).map {tuple in
			return tuple.second
		}
		objc_setAssociatedObject(self,__FUNCTION__, signal, objc_AssociationPolicy.OBJC_ASSOCIATION_RETAIN_NONATOMIC)
		return signal!
	}
}
